purchases/src/main/kotlin/com/revenuecat/purchases/Purchases.kt

```diff
@@ -1618,10 +1595,17 @@ class Purchases internal constructor(
             if (!appConfig.finishTransactions) {
                 log(LogIntent.WARNING, PurchaseStrings.PURCHASE_FINISH_TRANSACTION_FALSE)
             }
-            if (state.productChangeCallback == null) {
-                state = state.copy(productChangeCallback = listener)
-                userPurchasing = identityManager.currentAppUserID
+            if (callbackToUse is ProductChangeCallback && state.deprecatedProductChangeCallback == null) {
+                state = state.copy(deprecatedProductChangeCallback = deprecatedProductChangeCallback)
+            } else if (callbackToUse is PurchaseCallback) {
+                if (!state.purchaseCallbacksByProductId.containsKey(purchasingData.productId)) {
+                    val mapOfProductIdToListener = mapOf(purchasingData.productId to callbackToUse)
+                    state = state.copy(
+                        purchaseCallbacksByProductId = state.purchaseCallbacksByProductId + mapOfProductIdToListener
+                    )
+                }
```

Comment:
Does there need to be an `else` here at all? 🤔  I don't fully understand what's going on here but 🤷‍♂️

Reply:
this is adding a `PurchaseCallback` to a map keyed by productId, so that when the purchase returns, we know which callback to call. it was copy pasta from previous implementation.

but now that you ask, i know we have a `OperationAlreadyInProgress` error somewhere though, maybe it'd be best to call that here... 🤔 otherwise someone might think the callback will be replaced or added...

Reply:
actually, we do the `OperationAlreadyInProgress` further down in this method in this case 👍